### PR TITLE
feat: custom timeout for SubmitTX

### PR DESCRIPTION
**Pull request by pr0n00gler**
**Task**: https://p2pvalidator.atlassian.net/jira/software/projects/LSC/boards/121?selectedIssue=LSC-152

Custom timeout when sending Interchain Tx is implemented.

**DEPENDS ON**: https://github.com/neutron-org/neutron-contracts/pull/20

### CODE REVIEW

**Comment by foxpy** [comment]
Rebased on `feat/stateful-relayer-env` because it didn't work otherwise

**Review by foxpy — DISMISSED** [review]
Tested, everything works fine, code looks great

**Review by zavgorodnii — DISMISSED** [review]

File: x/interchaintxs/types/tx.go [hunk 1]
```diff
@@ -67,6 +67,10 @@ func (m *MsgSubmitTx) ValidateBasic() error {
 		return errors.New("no messages provided")
 	}
 
+	if m.Timeout == 0 {
+		return errors.New("timeout cannot be zero")
```
Comment (sotnikov-s) [inline x/interchaintxs/types/tx.go]: let's make it only positive greater than 0

File: proto/interchaintxs/v1/tx.proto [hunk 2]
```diff
@@ -55,6 +55,9 @@ message MsgSubmitTx {
   string connection_id = 3;
   repeated google.protobuf.Any msgs = 4;
   string memo = 5;
+
```
Comment (sotnikov-s) [inline proto/interchaintxs/v1/tx.proto]: y empty line?

File: wasmbinding/bindings/msg.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ type SubmitTx struct {
 	InterchainAccountId string        `json:"interchain_account_id"`
 	Msgs                []ProtobufAny `json:"msgs"`
 	Memo                string        `json:"memo"`
+	Timeout             int64         `json:"timeout"`
```
Comment (sotnikov-s) [inline wasmbinding/bindings/msg.go]: reasonably to have it as uint

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: x/interchaintxs/types/tx.go (hunk 1 above)
Comment (mmikev) [reply x/interchaintxs/types/tx.go]: Done

File: proto/interchaintxs/v1/tx.proto (hunk 2 above)
Comment (mmikev) [reply proto/interchaintxs/v1/tx.proto]: Done

File: wasmbinding/bindings/msg.go (hunk 3 above)
Comment (mmikev) [reply wasmbinding/bindings/msg.go]: Done

**Review by sotnikov-s — DISMISSED** [review]

File: x/interchaintxs/keeper/msg_server.go [hunk 4]
```diff
@@ -119,7 +119,7 @@ func (k Keeper) SubmitTx(goCtx context.Context, msg *ictxtypes.MsgSubmitTx) (*ic
 		Memo: msg.Memo,
 	}
 
-	timeoutTimestamp := time.Now().Add(InterchainTxTimeout).UnixNano()
+	timeoutTimestamp := ctx.BlockTime().Add(time.Duration(msg.Timeout) * time.Second).UnixNano()
```
Comment (swelf19) [inline x/interchaintxs/keeper/msg_server.go:119]: `InterchainTxTimeout` is a dead const now.

File: wasmbinding/test/custom_message_test.go [hunk 5]
```diff
@@ -239,14 +241,16 @@ func (suite *CustomMessengerTestSuite) TestSubmitTx() {
 		"connection_id": "%s",
 		"interchain_account_id": "%s",
 		"msgs": %s,
-		"memo": "%s"
+		"memo": "%s",
+		"timeout": %s
```
Comment (swelf19) [inline wasmbinding/test/custom_message_test.go]: `%s` is a bit confusing at this place?

**Review by swelf19 — DISMISSED** [review]

File: x/interchaintxs/keeper/msg_server.go (hunk 4 above)
Comment (mmikev) [reply x/interchaintxs/keeper/msg_server.go:119]: fixed

File: wasmbinding/test/custom_message_test.go (hunk 5 above)
Comment (mmikev) [reply wasmbinding/test/custom_message_test.go]: fixed

File: x/interchaintxs/types/tx.go (hunk 1 above)
Comment (swelf19) [reply x/interchaintxs/types/tx.go]: Actually since Timeout uint64 type it cannot be less than zero

**Review by swelf19 — APPROVED** [review]

**Review by zavgorodnii — APPROVED** [review]

**Review by foxpy — APPROVED** [review]

**Pull request merged**